### PR TITLE
Initial functionality for Eames (straight) vacuum brake.

### PR DESCRIPTION
**Pull request by peternewell**
https://blueprints.launchpad.net/or/+spec/vacuum-brake-update - adding Eames brake functionality

### CODE REVIEW

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs
Comment (YoRyan) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs:69]: Please note that we prefer to use 'var' whenever possible, especially in brand new code.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs
Comment (YoRyan) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs:299]: See my prior note about triggering sound events on value changes.

File: Source/Orts.Simulation/Common/Scripting/BrakeController.cs
Comment (peternewell) [reply Source/Orts.Simulation/Common/Scripting/BrakeController.cs:286]: The main reason was to save some screen space, and not make lines too long to read. 

Perhaps we need a standard approach to how names are used.

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs:2108]: Can you give an example of how you think that this would be called in the relevant code sections?

The only other thing that I am uncertain about is that it is a parameter that is "inherited" from vacuum brake class, so I am not sure whether this will create any problems.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs:72]: Can you be a bit more specific, and provide an example?

File: Source/Orts.Simulation/Common/Scripting/BrakeController.cs
Comment (YoRyan) [reply Source/Orts.Simulation/Common/Scripting/BrakeController.cs:286]: Having taken a second look I see a few other abbreviations, so it's not like you're doing anything novel. My mistake.

And yet we also have long strings like `Emergency Braking Push Button`, so maybe a standardized approach is warranted...

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs
Comment (YoRyan) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs:72]:
```suggestion
            var lead = (MSTSLocomotive)Car.Train.LeadLocomotive;
```

And

```var vr = NumBrakeCylinders * BrakeCylVolM3 / BrakePipeVolumeM3;```

not

```float vr = NumBrakeCylinders * BrakeCylVolM3 / BrakePipeVolumeM3;```

etc.

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs
Comment (YoRyan) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs:2108]: See the attached patch.
[ryan-vacuumbrake.diff.txt](https://github.com/openrails/openrails/files/5960518/ryan-vacuumbrake.diff.txt)

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs:72]: What is the difference between var and float, and why is it better to use it?

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs:2108]: Thanks for the patch file example.

I can't use the bool function **LargeSteamEjectorIsOn** because it is used is some code instances where the sound trigger is not required, so it can't be mixed.

Instead I have created a get/set function using as follows:

`        private bool _LargeEjectorSoundOn = false;
        public bool LargeEjectorSoundOn
        {
            get => _LargeEjectorSoundOn;
            private set
            {
                if (value != _LargeEjectorSoundOn)
                {
                    SignalEvent(value? Event.LargeEjectorOn : Event.LargeEjectorOff);
                    _LargeEjectorSoundOn = value;
                }
            }
       }   `

I have inserted this get/set parameter ok in the MSTSLocomotive.cs file, but when I attempt to use it in the StraightVacuumSinglePipe.cs file, I get an error which suggests that the "**set accessor is not accessible**". It allows me to select the parameter, but not compile with it.

Do you have any advice to overcome this issue?

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs]: Changed as requested

File: Source/Documentation/Manual/physics.rst
Comment (peternewell) [reply Source/Documentation/Manual/physics.rst:3039]: Thanks for picking that up, corrected

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs
Comment (YoRyan) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs:2108]: `private set` might be too restrictive. If you need to access that property in an inheriting class, try `protected`. Last resort is to use `public`.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs
Comment (YoRyan) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs]: `var` allows C# to automatically determine the type (`float`, `int`, `bool`, `MSTSLocomotive`, etc.) of variable using a technique called "type inference." We prefer the `var` keyword because compared to declaring an explicit type, it is briefer and less redundant (observe that you typed `MSTSLocomotive` twice in the current code).

Going back and changing all of your variable declarations would entail a lot of work, so I'm not suggesting you do that--but do try to prefer `var` in future submissions. Note that it is not always possible to do so--such as when you're declaring a variable without an initial value, for example.

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs:2108]: I have tried some different combinations of `protected` and `public`, and still doesn't work.

I have attached the current patch file.

[sound_trigger.diff.txt](https://github.com/openrails/openrails/files/5968452/sound_trigger.diff.txt)

Comment (YoRyan) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs:2108]: If you change the access levels of the setters to public, by dropping "private" from `private set` and just using `set`, both should work.

Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs:2108]: I had already tried changing the access level to public, but that caused a level clash.

I have removed private from the set statement, and that appears to work.

So I will upload some modifications.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs
Comment (cesarBLG) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs]: Ryan,
I agree with you that in `var loco = (MSTSLocomotive)Locomotive` the implicit type is shorter and more readable, as well as for `var text = "someText"` or `var pi = 3.1415f`.

However, for the expression `float vr = NumBrakeCylinders * BrakeCylVolM3 / BrakePipeVolumeM3;` and, in general, when the numeric type being used is not clear from the expression, I would prefer to explicitly set the type of the variable. I think it improves the readability and code analysis, because the explicit declaration ensures that you are using a 'float' and not a 'double' or an 'int'.

What's your opinion about it?

Comment (YoRyan) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs]: To be honest, I agree, and I use explicit typing in my personal code. @twpol told me about our project's preference for `var` in one of my early PR's. Perhaps he could clarify?

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs]: code modified

**Review by YoRyan — COMMENTED** [review]
No more objections from me, Peter. Since these are new features, however, I believe we need the approval of @cjakeman to merge.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs
Comment (cjakeman) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/StraightVacuumSinglePipe.cs]: When James explained this policy to me, I believe he agreed with Cesar:

> in general, when the numeric type being used is not clear from the expression, I would prefer to explicitly set the type of the variable.

**Review by cjakeman — APPROVED** [review]
Hi Peter,

All looks good. I note that the linked forum thread runs all the way from July 2019 !

Approved with thanks (and for all the suggested improvements)

Chris

**Pull request merged**
